File: client/tws_foundation_client/lib/src/models/exception_info.dart
import 'package:csm_client/csm_client.dart';
import 'package:tws_foundation_client/tws_foundation_client.dart';

/// Data {class} implementation for an [ExceptionInfo].
///
///
/// Defines a data constract for a [FoundationServer] implementation exception, storing diagnosticable exception information.
final class ExceptionInfo implements DecodableI {
  /// Where the exception got thrown ([ServerI] side).
  String trace = '';

  /// Specific exception management situation code.
  int situation = 0;

  /// User friendly advise message to display (recommended).
  String advise = '';

  /// A system reflected type with message from the exception thrown at server context.
  String system = '';

  /// A custom collection of data, this is custom per exception definition.
  DataMap factors = <String, Object?>{};

  /// Generates a new [ExceptionInfo] object.
  ExceptionInfo();

  @override
  void decode(DataMap encode) {
    situation = encode.get('situation');
    trace = encode.get('trace');
    advise = encode.get('advise');
    system = encode.get('system');
    factors = encode.get('factors');
  }
}

File: server/TWS Foundation/CSM Foundation/Server/ExceptionInfo.cs
namespace CSM_Foundation.Server;

/// <summary>
///     {model} record for <see cref="ExceptionInfo"/>.
///     
///     <para>
///         Stores public information to expose an Exception details.
///     </para>
/// </summary>
public record ExceptionInfo {
    /// <summary>
    ///     Where the exception got thrown.
    /// </summary>
    public required string Trace { get; init; }

    /// <summary>
    ///     Specific exception management situation code.
    /// </summary>
    public required int Situation { get; init; }

    /// <summary>
    ///     User friendly displayable error message.
    /// </summary>
    public required string Advise { get; init; }

    /// <summary>
    ///     Exception message caught at system level.
    /// </summary>
    public required string System { get; init; }

    /// <summary>
    ///     Additional detailed information about the exception.
    /// </summary>
    public Dictionary<string, dynamic> Factors { get; init; } = [];
}
